src/fuzz/protobufs/utils/spanner_emulator_ddl_statement_proto_to_string.cc

```diff
@@ -0,0 +1,188 @@
+//
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "src/fuzz/protobufs/create_table.pb.h"
+#include "src/fuzz/protobufs/spanner_ddl.pb.h"
+
+#include <google/protobuf/repeated_field.h>
+
+#include <cmath>
+#include <string>
+#include "absl/strings/substitute.h"
+#include "absl/strings/str_cat.h"
+
+using spanner_ddl::SpannerDDLStatement;
+using spanner_ddl::CreateTable;
+using spanner_ddl::Column;
+using spanner_ddl::ColumnDataInfo;
+
+// forward declarations
+std::string toString(SpannerDDLStatement statement);
+std::string toString(CreateTable createTable);
+std::string toString(const google::protobuf::RepeatedPtrField<Column> columns);
+std::string toString(Column column);
+std::string toString(ColumnDataInfo columnDataInfo);
+std::string toString(ColumnDataInfo::ScalarType type, 
+    int length, ColumnDataInfo::LengthType lengthType);
+std::string isColumnNotNullToString(bool isNotNull);
+std::string columnOptionsToString(bool allowCommitTimestamps);
+std::string toPrimaryKeys(const google::protobuf::RepeatedPtrField<Column>& columns);
+std::string columnToPrimaryKey(Column column);
+std::string toString(Column::Orientation orientation);
+
+// Transforms any Emulator DDL statement into a syntactically valid string
+std::string toString(SpannerDDLStatement statement) {
+    using statementType = SpannerDDLStatement::DDLStatementCase;
+    switch (statement.DDLStatement_case()) {
+        case statementType::kCreateTable:
+            return toString(statement.createtable());
+        //TODO: add more cases here for additional APIs
+        default:
+            return "";
+    }
+}
+
+// generates a 'CREATE TABLE ...' statement
+std::string toString(CreateTable createTable) {
+    return absl::Substitute("CREATE TABLE $0 ([ $1,$2 ]) PRIMARY KEY ( [$3] )",
+                createTable.tablename(), 
+                toString(createTable.primarykeys()),
```

Comment:
How does this behave if primary keys/other keys are empty?  Do we still get mostly valid DDL?

Reply:
(Also--we seem to remove the trailing comma from the column list--do we need to put one back here?)

Reply:
Added new method that checks if both primary and non-primary keys exist before joining them with commas. If both are non-existent then an empty string is returned instead